adventuregamestudio/ags
# Editor: Android build feature

This PR should be working now. Additional information in the [forum thread](https://www.adventuregamestudio.co.uk/forums/index.php?topic=59772.0).

https://user-images.githubusercontent.com/2244442/150422308-c7d3c43f-6dcd-45c0-ac7c-7a068ccf7b6e.mp4

## Test instructions

1. Load the AGS Editor built from this codebase and on general settings you should see an Android build option
  a. If you see no Android build option, then you are missing the android bundle. 
  b. If you are building from source you need to download the android bundle, it's named as `AGS-3.6.0.16-android-proj-release.zip` and you can get it from the CI artifacts. It's just the single game project, with the libraries put under libs, and it should be placed in a directory named `Android/` in your Editor root folder.
2. Install Android Studio, preferably the last one, and set the following environment variable in your Windows
  a.  `ANDROID_HOME`, set it to where you installed the Android SDK (usually `C:\Users\USERNAME\AppData\Local\Android\Sdk`)
  b. `JAVA_HOME`, set it to the jre in the Android Studio installation (usually in `C:\Program Files\Android\Android Studio\jre`)
3. Reboot, or log out, just make sure that these are properly applied
4. Now load the AGS Editor again and hit build EXE

---

Leaving in draft since this has a lot of things missing, but just get things started here. Here are some things to do:

- [x] figure how to put keystore info in preferences or **people are going to push their key passwords to git** (maybe someone will solve #1477 ?)
- [x] Allow setting where the keystore is through the Editor (in preferences), this is now set in `Compiled\Android\mygame\local.static.properties` 
- [x] Actually apply the package name, this can be done through gradle nowadays (no renaming)
- [x] move the .aab or .apk file after it's built to root of Android dir.
- [x] Set whether to package game inside apk or in asset dir with .aab
- [x] figure a way to pass file filters through Property decorators..
- [x] CI to prepare the package I made above
- [x] Allow configuring ANDROID_HOME and JAVA_HOME through Editor preferences.
- [x] Check with `sdkmanager` that dependencies (android platform sdk and ndk for the versions used) are in place, and if not, ask to install them (not sure if needed, maybe gradle can request them)
- [x] Override back button for keyboard toggle and extra keys (fix #1472)
- [x] Write `android.cfg` file
- [x] Properly write the app name used in Android launcher
- [x] Add Keystore Generation
- [x] APP icons are picked up from `icons/android/` dir at the AGS Project root, which has to contain android mipmap icon dirs with `ic_launcher.png` files.
- [x] automatically close the gradlew log window (but keep it if there's an error)
- [x] do not use gradle daemon by default, but allow turning it on through preference to speed up building

![image](https://user-images.githubusercontent.com/2244442/150614479-bddeda14-28f5-4a0a-b690-a262682938d1.png)

---

## FAQ

**Q:** Can I not install Android Studio?
**A:** I am using the exactly same tools from Android Studio - it requires installing Android Studio. You could theoretically install only Android SDK, Bundle Tool, Java JRE and JDK but then you need to piece everything together and it will take roughly the same storage space, so it's not worth it.

---

Fix #1337
